### PR TITLE
[WIP] Bi-directional streaming region heartbeats 

### PR DESCRIPTION
**Pull request by overvenus**
This PR replaces unary region heartbeats with bi-directional streaming heartbeats, it intends to increase heartbeats throughput and enable server push in the future.

Pending:

 - [ ] pingcap/kvproto#152
 - [ ] pingcap/pd#613
 - [x] pingcap/grpc-rust#3
 - [ ] #1800 switch to [grpc-rs](https://github.com/pingcap/grpc-rs)

### CODE REVIEW

File: src/util/transport.rs [hunk 1]
```diff
@@ -149,6 +152,19 @@ impl<T, C: Sender<T>> Clone for RetryableSendCh<T, C> {
     }
 }
 
+pub struct SendChStream<T, C: Sender<T>> {
```
Comment (siddontang) [inline src/util/transport.rs]: is it SendStreamCh? or StreamSendCh?

File: src/pd/async/client.rs [hunk 2]
```diff
@@ -346,34 +347,44 @@ impl PdClient for RpcClient {
...
-                .map_err(Error::Grpc)
-                .and_then(|resp| {
-                    try!(check_resp_header(resp.get_header()));
-                    Ok(resp)
-                })
-                .boxed()
-        };
+    fn region_heartbeat<S, E>(&self, req_stream: S) -> PdStream<pdpb::RegionHeartbeatResponse>
+        where E: Send + 'static,
+              S: Stream<Item = Option<(metapb::Region,
```
Comment (siddontang) [inline src/pd/async/client.rs]: why using tuple but not a structure here?

File: src/util/transport.rs (hunk 1 above)
Comment (overvenus) [inline src/util/transport.rs]: It's a stream of `SendCh`, `Stream::Item = SendCh`

File: src/raftstore/store/worker/pd.rs [hunk 3]
```diff
@@ -131,62 +139,77 @@ impl<T: PdClient> Runner<T> {
         handle.spawn(f)
     }
 
-    fn handle_heartbeat(&self,
+    fn handle_heartbeat(&mut self,
```
Comment (siddontang) [inline src/raftstore/store/worker/pd.rs]: why not using RegionHeartbeat here?

**Pull request closed without merging**